erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -0,0 +1,824 @@
+    Uint trap_data_size = sizeof(TrapData) / sizeof(Eterm);
+    Eterm* hp = HAlloc(BIF_P, trap_data_size);
+    TrapData* trap_data;
+    Eterm res = NIL;
+
+    trap_data = (TrapData *) hp;
+    trap_data->header = make_pos_bignum_header(trap_data_size-1);
+    trap_data->table = hash_table;
+    trap_data->idx = 0;
+    trap_data->remaining = hash_table->num_entries;
+    trap_data->memory = 0;
+    res = do_info(BIF_P, trap_data);
+    if (trap_data->remaining == 0) {
+        BIF_RET(res);
+    } else {
+        /*
+         * Increment the ref counter to prevent an update operation (by put/2
+         * or erase/1) to delete this hash table.
+         */
+        erts_atomic_inc_nob(&hash_table->refc);
+        BIF_TRAP2(&persistent_term_info_export, BIF_P, make_boxed(hp), res);
+    }
+}
+
+Uint
+erts_persistent_term_count(void)
+{
+    HashTable* hash_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    return hash_table->num_entries;
+}
+
+void
+erts_init_persistent_dumping(void)
+{
+    HashTable* hash_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    ErtsLiteralArea** area_p;
+    Uint i;
+
+    /*
+     * Overwrite the array of Eterms in the current hash table
+     * with pointers to literal areas.
+     */
+
+    erts_persistent_areas = (ErtsLiteralArea **) hash_table->term;
+    erts_num_persistent_areas = hash_table->num_entries;
+    area_p = erts_persistent_areas;
+    for (i = 0; i < hash_table->allocated; i++) {
+        Eterm term = hash_table->term[i];
+
+        if (is_boxed(term)) {
+            *area_p++ = term_to_area(term);
+        }
+    }
+}
+
+/*
+ * Local functions.
+ */
+
+static HashTable*
+create_initial_table(void)
+{
+    HashTable* hash_table;
+    int i;
+
+    hash_table = (HashTable *) erts_alloc(ERTS_ALC_T_PERSISTENT_TERM,
+                                          sizeof(HashTable)+sizeof(Eterm) *
+                                          (INITIAL_SIZE-1));
+    hash_table->allocated = INITIAL_SIZE;
+    hash_table->num_entries = 0;
+    hash_table->mask = INITIAL_SIZE-1;
+    hash_table->first_to_delete = 0;
+    hash_table->num_to_delete = 0;
+    erts_atomic_init_nob(&hash_table->refc, (erts_aint_t)1);
+    for (i = 0; i < INITIAL_SIZE; i++) {
+        hash_table->term[i] = NIL;
+    }
+    return hash_table;
+}
+
+static BIF_RETTYPE
+persistent_term_get_all_trap(BIF_ALIST_2)
+{
+    TrapData* trap_data = (TrapData *) BIF_ARG_1;
+    Eterm res = BIF_ARG_2;
+
+    ASSERT(is_big(BIF_ARG_1));
+    ASSERT(is_list(BIF_ARG_2));
+    trap_data = (TrapData *) boxed_val(BIF_ARG_1);
+    res = do_get_all(BIF_P, trap_data, res);
+    ASSERT(is_list(res));
+    if (trap_data->remaining > 0) {
+        BUMP_ALL_REDS(BIF_P);
+        BIF_TRAP2(&persistent_term_get_all_export, BIF_P, BIF_ARG_1, res);
+    } else {
+        /*
+         * Decrement ref count (and possibly delete the hash table
+         * and associated literal area).
+         */
```

Comment:
No reds bumped when we've gotten all the data?

Reply:
Will fix.